tools/testing/e2e/iree-e2e-fa2-test.cc

```diff
@@ -0,0 +1,471 @@
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "iree/base/api.h"
+#include "iree/base/internal/cpu.h"
+#include "iree/base/internal/flags.h"
+#include "iree/base/internal/math.h"
+#include "iree/base/internal/path.h"
+#include "iree/hal/api.h"
+#include "iree/modules/hal/module.h"
+#include "iree/tooling/context_util.h"
+#include "iree/tooling/device_util.h"
+#include "iree/vm/api.h"
+#include "iree/vm/native_module_cc.h"
+#include "tools/testing/e2e/test_utils.h"
+
+//===----------------------------------------------------------------------===//
+// Reference Attention
+//===----------------------------------------------------------------------===//
+
+// Helper for reference_attention.
+// Function to allocate and initialize tensors
+float* allocate_tensor(int dim1, int dim2, int dim3) {
+  float* tensor = (float*)malloc(dim1 * dim2 * dim3 * sizeof(float));
+  for (int i = 0; i < dim1 * dim2 * dim3; ++i) {
+    tensor[i] = 0.0f;
+  }
+  return tensor;
+}
+
+// Function to free allocated tensors
+void free_tensor(float* tensor) { free(tensor); }
+
+// Function to calculate 1D index for a 3D array
+int index_3d(int i, int j, int k, int dim2, int dim3) {
+  return i * dim2 * dim3 + j * dim3 + k;
+}
+
+static void reference_attention_f16_f16_f16_f16(
+    iree_hal_dim_t M, iree_hal_dim_t K1, iree_hal_dim_t K2, iree_hal_dim_t N,
+    iree_hal_dim_t B, const float* query_data, const float* key_data,
+    const float* value_data, float* result_data, iree_hal_dim_t b) {
+  float* Attention = allocate_tensor(B, M, K2);
+
+  // Compute Q * K^T
+  for (int m = 0; m < M; ++m) {
+    for (int k2 = 0; k2 < K2; ++k2) {
+      float sum = 0.0;
+      for (int k1 = 0; k1 < K1; ++k1) {
+        int q_idx = index_3d(b, m, k1, M, K1);
+        int k_idx = index_3d(b, k2, k1, K2, K1);
+        sum += query_data[q_idx] * key_data[k_idx];
+      }
+      int att_idx = index_3d(b, m, k2, M, K2);
+      Attention[att_idx] = sum / sqrt(K1);  // Scale by sqrt(K1)
+    }
+  }
+
+  // Compute softmax on Attention
+  for (int m = 0; m < M; ++m) {
+    // Calculate softmax denominator
+    float sum = 0.0;
+    for (int k2 = 0; k2 < K2; ++k2) {
+      int att_idx = index_3d(b, m, k2, M, K2);
+      sum += exp(Attention[att_idx]);
+    }
+    // Apply softmax
+    for (int k2 = 0; k2 < K2; ++k2) {
+      int att_idx = index_3d(b, m, k2, M, K2);
+      Attention[att_idx] = exp(Attention[att_idx]) / sum;
+    }
+  }
+
+  // Compute Attention * V
+  for (int m = 0; m < M; ++m) {
+    for (int n = 0; n < N; ++n) {
+      float sum = 0.0;
+      for (int k2 = 0; k2 < K2; ++k2) {
+        int att_idx = index_3d(b, m, k2, M, K2);
+        int v_idx = index_3d(b, k2, n, K2, N);
+        sum += Attention[att_idx] * value_data[v_idx];
+      }
+      int o_idx = index_3d(b, m, n, M, N);
+      result_data[o_idx] = sum;
+    }
+  }
+}
+
+static iree_status_t reference_attention_element(
```

Comment:
Please add an implementation for the f32 type as well. It might be good to double-check numerics.

Reply:
I can add that too. Can we make a comparison with the two references?